### PR TITLE
Add benchmark tests

### PR DESCRIPTION
**Pull request by jalvz**

### CODE REVIEW

File: errs/errs.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package errs
+
+func FirstNotNil(errs ...error) error {
```
[the remark this answers is not in the document]
Comment (simitt) [reply errs/errs.go]: For me this is not idiomatic to read. The way you use it [here](https://github.com/elastic/hey-apm/pull/116/files#diff-ea52d2e720f7b23b44d31c84517d3b84R49) does actually mean that potentially many unnecessary calls are made. 

According to your comment you are fully aware of that, so your call.

File: benchmark_test/regression.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+			"range": types.M{
+				"@timestamp": types.M{
+					"gte": fmt.Sprintf("now-%sd/d", days),
+					"lt":  "now",
+				},
+			},
+		},
+	}
+	for k, v := range inputMap {
+		filters = append(filters, types.M{"match": types.M{k: v}})
```
Comment (simitt) [inline benchmark_test/regression.go]: Does this mean that if an `APMServerUrl`, `APMServerSecret` etc. changes the results will no longer be comparable?

File: es/api.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	Url      string
+	username string
+	password string
+}
+
+// returns a client connected to an ElasticSearch node with given `params`
+// "local" is short for http://localhost:9200
+func NewConnection(url, auth string) (Connection, error) {
+	if url == "local" {
+		url = "http://localhost:9200"
```
Comment (simitt) [inline es/api.go]: Could also be a const.

File: benchmark_test/regression.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+				"must": filters,
+			},
+		},
+	}
+
+	savedReports, fetchErr := es.FetchReports(conn, body)
+
+	var regression error
+	for _, sr := range savedReports {
+		if r.ReportId != sr.ReportId && r.Performance()*margin < sr.Performance() {
```
Comment (simitt) [inline benchmark_test/regression.go]: I haven't given this a lot of thoughts yet, but that basically means that if each of n changes over the last couple of days has decreased performance a bit, all of them are fine. But at some point one change is not within the margin boundary compared to the oldest fetched report, and then a regression is raised, although this specific change might not be the general problem.  
I was wondering if you've looked into alternatives or how the agent devs check this?

File: es/api.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	)
+
+	var respErr error
+	if resp.IsError() {
+		respErr = errors.New(resp.String())
+	}
+
+	parsed := SearchResult{}
+	decodingErr := json.NewDecoder(resp.Body).Decode(&parsed)
+
```
Comment (simitt) [inline es/api.go]: If you'd create `ret` here you would already know the size.

File: models/report.go [hunk 6]
```diff
@@ -147,7 +139,14 @@ func (r Report) date() time.Time {
 	return t
 }
 
-func WithDerivedAttributes(r Report) Report {
+func (r Report) Performance() float64 {
+	if r.EventIndexRate != nil {
+		return *r.EventIndexRate
+	}
+	return 0
```
Comment (simitt) [inline models/report.go]: Could `EventIndexRate` be a `float64` instead of a `*float64`?

File: benchmark_test/benchmark_test.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+	"github.com/pkg/errors"
+
+	"go.elastic.co/apm"
+
+	"github.com/elastic/hey-apm/conv"
+
+	"github.com/stretchr/testify/assert"
+
+	. "github.com/elastic/hey-apm/benchmark_test"
```
[the remark this answers is not in the document]
Comment (jalvz) [reply benchmark_test/benchmark_test.go]: not necessary, but convenient to not have to qualify helper functions in the same package 
https://golang.org/ref/spec#Import_declarations

File: benchmark_test/benchmark_test.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
+package benchmark_test
```
[the remark this answers is not in the document]
Comment (jalvz) [reply benchmark_test/benchmark_test.go]: first: all login in `main` is reused here (just moved out of `main` which is a good idea any case)

second: jenkins will run this as `ENV_VAR_X=X... go test -v ....`, which means that to add a workload we just have to add a unit test; and you can run it locally in the same way
the other way would mean that jenkins would run tests like:
```
./hey-apm benchmark -run 3m -tf 100ms -e 100000 -mx 10 -ms 10 -ef 20 (...many more args)
./hey-apm benchmark -run 3m -tf 100ms -e 0 -mx 30 -ms 30 (...many more args)
...many more lines
```
adding more workloads means adding more lines to some script, which I don't know if I can run locally (jenkins files?), etc.; and it is much more unreadable (like 10 lines with 15 args each), easier to slip typos, etc.

tl;dr golang is more expressive than a make script or similar. i can change it if this has any drawbacks i didn't think about.

File: errs/errs.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jalvz) [reply errs/errs.go]: first, tests run for, say, 30 minutes.
second, we query elasticsearch (<1 second)
then we iterate over a loop and compare 2 values on each iteration (a few microsecs?)

so, in the case that elasticsearch query fails, i don't mind indeed adding a few microsecs to 30 minutes.

File: benchmark_test/benchmark_test.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+func TestTransactionsWithSpans(t *testing.T) {
+	report := worker.Run(builder().Input)
+	verify(t, report)
+}
+
+func TestTransactionsWithManySpans(t *testing.T) {
+	report := worker.Run(
+		builder().WithSpans(50).Input)
```
[the remark this answers is not in the document]
Comment (jalvz) [reply benchmark_test/benchmark_test.go]: good idea :+1:

File: benchmark_test/benchmark_test.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+	. "github.com/elastic/hey-apm/benchmark_test"
+	"github.com/elastic/hey-apm/es"
+	"github.com/elastic/hey-apm/models"
+	"github.com/elastic/hey-apm/worker"
+)
+
+var (
+	// global variables initialized in TestMain - use with care
+	glRegressionMargin float64
```
[the remark this answers is not in the document]
Comment (jalvz) [reply benchmark_test/benchmark_test.go]: idk, don't think so. is it wrong for not being a common pattern? is it anti-intuitive or complicated?

File: benchmark_test/benchmark_test.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+	input := builder().WithErrors().Input
+	input.RunTimeout = glWarmup
+	input.ElasticsearchUrl = ""
+	fmt.Println(fmt.Sprintf("warming up %.1f seconds...", glWarmup.Seconds()))
+	worker.Run(input)
+	coolDown()
+}
+
+// waits an arbitrary time for events in elasticsearch be flushed
+func coolDown() {
```
[the remark this answers is not in the document]
Comment (jalvz) [reply benchmark_test/benchmark_test.go]: there is a `flushTimeout` name, and a `flush` function somewhere else, i wanted to avoid ambiguity where it was easy to avoid. it is not just waiting for ES flush, but apm-server to free some memory, etc.

File: benchmark_test/builder.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+import (
+	"math"
+	"time"
+
+	"github.com/elastic/hey-apm/models"
+)
+
+type Wrap struct {
+	models.Input
+}
```
[the remark this answers is not in the document]
Comment (jalvz) [reply benchmark_test/builder.go]: > wrap should require no knowledge about how the input looks like.

why not? that is exactly its job, as `Input` is a direct field of `Wrap`. that said, looks like your grip is that the default Input is not the zero value for the type.
it is not for convenience; but that is an easy change, so will do that (basically get rid of the builder pattern)

File: benchmark_test/regression.go (hunk 2 above)
Comment (jalvz) [reply benchmark_test/regression.go]: `ApmServerUrl` yes (why wouldn't?, it would be exercising a different machine), `ApmServerSecret` no

File: benchmark_test/regression.go (hunk 4 above)
Comment (jalvz) [reply benchmark_test/regression.go]: > although this specific change might not be the general problem

correct, but you would look into it, inspect dashboards, commit history, etc. the important thing is that we get notified when performance decreases bellow a margin.

as far as i know, agent devs don't check for regressions at all (i might be wrong)

File: models/report.go (hunk 6 above)
Comment (jalvz) [reply models/report.go]: `EventIndexRate` is calculated from a division. If the denominator is zero, then this is nil (which is semantically different from being 0), so its needs a pointer

File: benchmark_test/benchmark_test.go [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+func TestHighLoad(t *testing.T) {
+	report := worker.Run(
+		builder().WithErrorFreq(time.Microsecond).WithTransactionFreq(time.Microsecond).Input)
+	verify(t, report)
+}
+
+// asserts there are no performance regressions for a given workload
+func verify(t *testing.T, report models.Report) {
+	coolDown()
+	assert.True(t, report.EventsIndexed > 100, fmt.Sprintf("not enough events indexed: %d", report.EventsIndexed))
```
[the remark this answers is not in the document]
Comment (jalvz) [reply benchmark_test/benchmark_test.go]: It is arbitrary, meaning that if there are less than 100 events indexed, then something definitely went wrong. 
I can change it for another arbitrary number, or remove the assertion altogether.

File: benchmark_test/builder.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (simitt) [reply benchmark_test/builder.go]: I did not mean to get rid of the builder pattern. My point is that the actual values of `input` could be completely different than defined in `TestMain`, e.g. `ErrorLimit: 100` and `TransactionLimit: 0`. In this case how would one create a wrap with no errors or with transactions? The struct `Wrap` holding a `model.Input` does not justify modeling the methods strongly coupled to one specific input.

File: benchmark_test/regression.go (hunk 2 above)
Comment (simitt) [reply benchmark_test/regression.go]: I missed that you only decode specific attributes of the struct, all good then 👍

File: benchmark_test/regression.go (hunk 4 above)
Comment (simitt) [reply benchmark_test/regression.go]: Fair enough. I was simply curious how others do it and if you have investigated other options, especially having in mind that we might want to be able to trigger this for specific PRs at some point.

File: benchmark_test/benchmark_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (simitt) [reply benchmark_test/benchmark_test.go]: I was referring to the already existing logic of [parsing user input](https://github.com/elastic/hey-apm/pull/116/files#diff-7ddfb3e035b42cd70649cc33393fe32cR16). Why introducing a second way to parse ENV variables instead of unifying and reusing?

I don't think the comparison of how you would start the tests and how you'd start a benchmark test is fair. You could still trigger _one_ `hey-apm benchmark` with the same options that are provided for starting the testsuite and then this `benchmark` task takes care of starting all the defined use cases. 

Not sure why you are referring to _make_ scripts, I didn't mean or imply to use any _make_ scripts. What I had in mind is something like:

```
struct Benchmark{
  args models.Input
  cases []models.Input
}

func (b *Benchmark) NewBenchmark(i models.Input){ 
  b.args = i
  //define and create cases based on i
}

func (b *Benchmark) Run(){
  for c := range b.cases{
    setup //warmup, cooldown etc
    report := worker.Run(c)
    if err := verify(report){ raise... }
  }
}
```

As I tried to explain in my first comment, I cannot follow your reasons for putting this logic into tests. This is a dedicated repository for running a high load against an external resource and gathering feedback about it. The benchmarking package adds logic to compare those results with past results. 
If you change it or not is up to you, but I'd appreciate if you could explain why you'd rather put this into tests than into regular go logic.

[the remark this answers is not in the document]
Comment (jalvz) [reply benchmark_test/benchmark_test.go]: Ok, so no make scripts, I misunderstood that, then.

So what you are arguing is that instead of `SERVER_URL=x... go test -v`, this should be executed like `go benchmark -server_url=x...`?

All the extra logic I am introducing is a `getEnv` function (not crazy complicated), and I gain access to all the test assertion functions, which are convenient. 
(and env vars might still be better either case for secrets, security wise)

Since this _tests_ the apm-server for regressions and will run on Jenkins, I also feel is quite natural to define it as tests.
So I still don't see clear downsides of this and advantages of the other, but is not big deal to change it... There are many ways to skin a cat, I guess.

File: benchmark_test/regression.go (hunk 4 above)
Comment (jalvz) [reply benchmark_test/regression.go]: Right now we have _nothing_, so if we get _something_, it is going to be better imo. 
If we find out that it is not good enough, we can iterate. 
Adjusting the number of days to look back and the error margin should also give some degree of flexibility.

File: benchmark_test/benchmark_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (simitt) [reply benchmark_test/benchmark_test.go]: My concerns are _conceptually_, not about how the functionality is called.

As you said, many different ways of solving something. Fine with me to leave it as is.

File: benchmark_test/benchmark_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (simitt) [reply benchmark_test/benchmark_test.go]: >  is it wrong for not being a common pattern?

didn't imply that, only asked if it is a pattern I don't know.

>  is it anti-intuitive or complicated?

It's not complicated. For me it is not super intuitive either, as `gl` for global is not a standardized abbreviation, and I am not used to this from any other language.

Keep it as is as you find it helpful.

File: benchmark_test/benchmark_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (simitt) [reply benchmark_test/benchmark_test.go]: This is my understanding so far:
* a package is not allowed to import itself. 
* if you want to use unexported funcationality use _whitebox testing_ by using the same package name in the `_test` file
* if you only want to do _blackbox testing_ use a package name with appendix `_test`

Trying to educate myself here, could you shed some light into why the package declares an import path of itself and why the directory containing the benchmarking logic is called `benchmark_test` and not `benchmark` although it also contains a `benchmark` package. 
Do you by any chance have a link or could explain what generally the advantage is of using a dedicated _test package but then importing the regular package to be able to also use unexported functions?

File: benchmark_test/benchmark_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (simitt) [reply benchmark_test/benchmark_test.go]: fair enough, wasn't aware of the cooling down of the server.

File: benchmark_test/benchmark_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (simitt) [reply benchmark_test/benchmark_test.go]: Mh, not sure about the implications of removing it. Are reports written to ES in case of performance decrease?

File: benchmark_test/regression.go (hunk 4 above)
Comment (simitt) [reply benchmark_test/regression.go]: Ofc that's all true, and it is a great start. I was only curious since I haven't looked into that at all.

File: benchmark_test/benchmark_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (simitt) [reply benchmark_test/benchmark_test.go]: Doesn't need to be addressed within this PR, but maybe create a follow up task.

File: benchmark_test/benchmark_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (simitt) [reply benchmark_test/benchmark_test.go]: While this is outdated now I'd still be interested in chatting about this if you're up for it.

File: benchmark_test/benchmark_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jalvz) [reply benchmark_test/benchmark_test.go]: this should be ok in this pr, i took some number from the doc
